### ISSUE TITLE
floopy_files returns 401

### ISSUE THREAD

**Issue by casz** [jetbrains-infra/packer-builder-vsphere#121]
```
==> vsphere-iso: Creating VM...
==> vsphere-iso: Customizing hardware...
==> vsphere-iso: Adding CD-ROM drives...
==> vsphere-iso: Creating floppy disk...
    vsphere-iso: Copying files flatly from floppy_files
    vsphere-iso: Done copying files from floppy_files
    vsphere-iso: Collecting paths from floppy_dirs
    vsphere-iso: Resulting paths from floppy_dirs : [C:\code\ThreeShape.Baseboxes/windows/setup/]
    vsphere-iso: Recursively copying : C:\code\ThreeShape.Baseboxes/windows/setup/
    vsphere-iso: Done copying paths from floppy_dirs
==> vsphere-iso: Uploading created floppy image
==> vsphere-iso: Destroying VM...
Build 'vsphere-iso' errored: 401 Unauthorized
```

If I remove the `floopy_dirs`/`floopy_files` definition I can get past it creation of VM

I have full rights on the datastore and yet seems impossible to upload.
Or perhaps it is the url that returns 401....
`https://vcenter.3shape.local/folder/example-windows/packer-tmp-created-floppy.flp?dcPath=%2F3Shape+Datacenter&dsName=Datastore1`

Debugging packer with `PACKER_LOG=1` provides nothing extra...

What permission am I missing? Do I need administrator on the datacenter?

Following the example to a T 😢
vSphere is v6.0

**Comment by casz** [comment]
Error messages seems weak and I have no experience setting up a debug environment for Go.

**Comment by casz** [comment]
Not sure the documentation
>floppy_files(array of strings) - List of local files to be mounted to the VM floppy drive. Can be used to make Debian preseed or RHEL kickstart files available to the VM.
floppy_img_path(string) - Data store path to a floppy image that will be mounted to the VM. Cannot be used with floppy_files or floppy_dir options. Example [datastore1] ISO/VMware Tools/10.2.0/pvscsi-Windows8.flp.

`Cannot be used with floppy_files or floppy_dir options.`

Yet the windows example shows exactly that example

https://github.com/jetbrains-infra/packer-builder-vsphere/blob/8754ddc709902e2f9575cc8e6f213542eda44ff8/examples/windows/windows-10.json#L35-L38

**Comment by schmandforke** [comment]
Hey, 

this issue can be caused by Untrusted Certificates.
If you try to upload an file through the webinterface, you should get the same error. The vSphere Server is redirecting the request to the ESXi Server in the backend which uses another selfsigned(?) cert. You have to trust this to upload the flp file.

**Comment by schmandforke** [comment]
can you even curl (without '-k') against the https endpoint from the same machine where packer running ?

**Comment by casz** [comment]
Definitely a certificate issue: `curl: (60) SSL certificate problem: unable to get local issuer certificate` Cheers for the hint 👍

**Comment by casz** [comment]
The issue is definitely permissions 👍 I am able to browse to `/folder` and see our data centers but no further 😭

**Comment by schmandforke** [comment]
Then doublecheck your account permissions... You have to be able to write in datastores.

**Comment by casz** [comment]
I am able to write to datastores I am just not able to browse our datacenter... coworker has the correct permissions.

**Comment by casz** [comment]
Closing in the end it was a permission thing, would be nice if the error message showed which URL it tried to hit.

**Comment by mkuzmin** [comment]
for file upload vSphere (or its golang API) requires weird permissions on top level:

* Datacenter (this object): `Datastore -> Low level file operations`
* Host (this object): `Host -> Configuration -> System Management`

**Comment by mkuzmin** [comment]
Another related issue: files are uploaded into a datastore through a host. By default a host is selected randomly. If an account has read-only permission over multiple hosts, they may be chosen for upload.

Fixed in e1cdd04f4f19a82b89bc6fe881261dc2c3197921.

**Comment by juliostanley** [comment]
@mkuzmin Sorry for commenting on a closed ticket, but I have this same exact issue, even while using packer 1.7.0, its while using granular permissions on a vcenter cluster, otherwise no issue on a different vcenter with non scoped permissions.

### The following have already been given:
 
> for file upload vSphere (or its golang API) requires weird permissions on top level:
> 
> * Datacenter (this object): `Datastore -> Low level file operations`
> * Host (this object): `Host -> Configuration -> System Management`

---

Using the same credentials, it is possible to upload files via the HTML 5 GUI, but using packer it fails.

I've noticed the path used by the HTML 5 GUI of vsphere, is different from the path used by vsphere-iso (checked through network tools the path). Both do a `PUT` to a similar url.

HTML 5 Client: PUT `https://<host>/folder/<foldername>/<somefilename>.flp?dcPath=ha-datacenter&dsName=DataStoreNameAsSeenOnVsphere`
vsphere-iso:  PUT `https://<host>/folder/<foldername>/<somefilename>.flp?dcPath=%2FDatacenterNameAsSeenOnVsphere&dsName=DataStoreNameAsSeenOnVsphere`

Notice that the vsphere html 5 client uses `dcPath=ha-datacenter` instead of `dcPath=%2FDatacenterNameAsSeenOnVsphere` not sure if that could be related.

---

- @mkuzmin Any thoughts?
- @jetersen What permissions corrected your scenario? Was it the same as described by @mkuzmin?
